gaia-header.js

```diff
@@ -34,124 +54,519 @@ module.exports = Component.register('gaia-header', {
+   * for DOM mutations.
+   *
+   * @private
+   */
+  observerStop: function() {
+    if (!this.observing) { return; }
+    this.observer.disconnect();
+    this.observing = false;
+    debug('observer stopped');
+  },
+
+  /**
+   * When the components DOM changes we run
+   * the debounced version runFontFit.
+   *
+   * @param  {Array} mutations
+   * @private
+   */
+  onMutation: function(mutations) {
+    debug('on mutation', mutations);
+    this.runFontFit();
+  },
+
+  /**
+   * Get the title width.
+   *
+   * Returns the space available for <h1>.
+   *
+   * @return {Number}
+   * @private
+   */
+  getTitleSpace: function() {
+    var titlePadding = 20;
+    var previous = this._titleSpace;
+    var value = this.getWidth()
+      - this.titleStart
+      - this.titleEnd
+      - titlePadding;
+    debug('get title space', value);
+    return value;
+  },
+
+  /**
+   * Get the width of the component.
+   *
+   * For performance reasons we make the
+   * assumption that the width is the same
+   * as `window.innerWidth` unless the
+   * `not-flush` attribute is used.
+   *
+   * @return {Number}
+   * @private
+   */
+  getWidth: function() {
+    var value = this.notFlush
+      ? this.clientWidth
+      : window.innerWidth;
+
+    debug('get width', value);
+    return value;
+  },
+
+  /**
+   * Triggers the 'action' button
+   * (used in Gaia integration testing).
+   *
+   * @public
+   */
+  triggerAction: function() {
+    if (this.action) { this.els.actionButton.click(); }
   },
 
   /**
    * Handle clicks on the action button.
    *
-   * Fired async to allow the 'click' event
+   * Fired pending to allow the 'click' event
    * to finish its event path before
    * dispatching the 'action' event.
    *
    * @param  {Event} e
    * @private
    */
-  onActionButtonClick: function(e) {
-    var config = { detail: { type: this.getAttribute('action') } };
-    var actionEvent = new CustomEvent('action', config);
-    setTimeout(this.dispatchEvent.bind(this, actionEvent));
+  onActionButtonClick: function() {
+    debug('action button click');
+    var config = { detail: { type: this.action } };
+    var e = new CustomEvent('action', config);
+    setTimeout(() => this.dispatchEvent(e));
+  },
+
+  /**
+   * Get the margin-start value required
+   * to center the title between
+   * surrounding buttons.
+   *
+   * @param  {Object} title
+   * @return {Object}
+   * @private
+   */
+  getTitleMarginStart: function() {
+    var start = this.titleStart;
+    var end = this.titleEnd;
+    var marginStart = end - start;
+    debug('get title margin start', marginStart);
+    return marginStart;
+  },
+
+  /**
+   * Get all the buttons (<a> & <button>)
+   * before the first <h1>.
+   *
+   * @return {Array}
+   * @private
+   */
+  getButtonsBeforeTitle: function() {
+    var children = this.children;
+    var l = children.length;
+    var els = [];
+
+    for (var i = 0; i < l; i++) {
+      var el = children[i];
```

Comment:
it's technically faster to iterate over children using firstElementChild (for initial state) + nextElementSibling (a iterator). But we'll have 2 buttons max so let's not focus on perf here, using "children" is more readable.